src/gc.rs

```diff
@@ -49,6 +51,18 @@ impl<T> Gc<T> {
     }
 }
 
+impl Gc<dyn Any> {
+    pub fn downcast<T: Any>(self) -> Result<Gc<T>, Gc<dyn Any>> {
```

Comment:
I'm not sure I see the utility in having `downcast` take `self` rather than `&self`. `downcast(self)` makes sense on a `Box`, for example, because you're also consuming the underlying memory, but on a `Gc` you're not. Thoughts?

Reply:
[For example, `Val::downcast` in yksom takes `&self` https://github.com/softdevteam/yksom/blob/master/src/lib/vm/val.rs#L130]

Reply:
I did think this, but I actually think it makes sense to consume the `Gc<T>` wrapper as we return a new one with the more concrete type. This feels a bit more in-keeping with Rust's move semantics. While it's true that we implement `Copy` anyway. It feels a bit more explicit and means we won't have handles of different types lingering in scope unless we really ask for it. 

This is also the `Rc` approach btw, so it also means we would have similar APIs.

Reply:
I know this is how `Rc` does it but I haven't yet really worked out *why* `Rc` does it this way (I guess that `Rc` followed `Box` and I can see why `Box` does it this way). I am unable to see any benefits from `self` (vs. `&self`) and some disadvantages (sometimes `downcast` forces you to `clone` a `Gc` in order that you still have a reference to the non-downcasted version). Put another way, everything you can do with `self` you can do with `&self` but not vice versa AFAICS?

Reply:
Yes that's right. I don't have very strong opinions on this, although I would personally lean further towards consistency. If you think it would make life easier as a consumer of `gcmalloc` for `Gc` to use `&self` then I don't mind changing it.

Reply:
Consistency is a laudable goal, but sometimes being inconsistent is more laudable :) Since we agree that `&self` is strictly more flexible -- and doesn't have any downsides that we know of -- I think that's the way to go.

Reply:
Fixed: 70f15f73309e0c26cf3ea58cc597f6d3570e5143